src/csharp/Sessionless/Impl/Sessionless.cs

```diff
@@ -0,0 +1,108 @@
+    }
+    public KeyPairHex GenerateKeys() {
+        var pair = KeyUtils.GenerateKeyPair();
+        return StoreHexPair(pair);
+    }
+
+    public async Task<KeyPairHex> GenerateKeysAsync() {
+        var pair = await KeyUtils.GenerateKeyPairAsync();
+        return StoreHexPair(pair);
+    }
+
+    private KeyPairHex StoreHexPair(AsymmetricCipherKeyPair pair) {
+        var simple = pair.ToHex();
+        Vault.Save(simple);
+        return simple;
+    }
+
+
+    public KeyPairHex? GetKeys() {
+        return Vault.Get();
+    }
+
+    public MessageSignatureHex Sign(string message) {
+        var privateHex = GetKeys()?.PrivateKey
+            ?? throw new KeyPairNotFoundException();
+        return Sign(message, privateHex);
+    }
+
+    public MessageSignatureHex Sign(string message, string privateKeyHex) {
+        // private hex to bigint
+        var privateInt = new BigInteger(privateKeyHex, 16);
+        // secp256k1 spec curve
+        ECDomainParameters curve = KeyUtils.Defaults.DomainParameters;
+        var privateKey = new ECPrivateKeyParameters(privateInt, curve);
+        return Sign(message, privateKey);
+    }
+
+    public MessageSignatureHex Sign(string message, ECPrivateKeyParameters privateKey) {
+        var paramWithRandom = new ParametersWithRandom(privateKey, new SecureRandom());
+        // init signer
+        var signer = new ECDsaSigner(new HMacDsaKCalculator(new Sha256Digest()));
+        signer.Init(true, paramWithRandom);
+        // message string to keccak256 hash
+        byte[] messageHash = message.HashKeccak256();
+        // sign
+        BigInteger[] signature = signer.GenerateSignature(messageHash);
+        return new MessageSignatureInt(signature).ToHex();
+    }
+
+
+    public bool Verify(SignedMessage signedMessage) {
+        var publicHex = GetKeys()?.PublicKey
+            ?? throw new KeyPairNotFoundException();
+        if (publicHex != signedMessage.PublicKey) return false;
+        return Verify(signedMessage, publicHex);
+    }
+    public bool Verify(SignedMessage signedMessage, string publicKeyHex) {
+        if (!publicKeyHex.IsHex()) {
+            throw new HexFormatRequiredException(nameof(publicKeyHex));
+        }
+        if (publicKeyHex != signedMessage.PublicKey) return false;
+        // public hex to bytes
+        var publicBytes = Hex.Decode(publicKeyHex);
+        // public bytes to key object
+        ECDomainParameters curve = KeyUtils.Defaults.DomainParameters;
+        ECPoint qPoint = curve.Curve.DecodePoint(publicBytes);
+        var publicKey = new ECPublicKeyParameters(qPoint, curve);
+        return Verify(signedMessage, publicKey);
+    }
+    public bool Verify(SignedMessage signedMessage, ECPublicKeyParameters publicKey) {
+        // signature hex to bigint (and hex still included)
+        var signature = signedMessage.Signature.ToInt();
+        // message string to keccak256 hash
+        byte[] messageHash = signedMessage.Message.HashKeccak256();
```

Comment:
we only do the first 32 bytes of the messageHash

Reply:
`HashKeccak256` is an extension method which produces 32 bytes
### Full method:
```csharp
    /// <summary> Get the hash of <paramref name="input"/> in keccak256 format </summary>
    internal static byte[] HashKeccak256(this string input) {
        // input to bytes
        byte[] bytes = Encoding.UTF8.GetBytes(input);
        //init digest
        KeccakDigest keccak256 = new(256);
        keccak256.BlockUpdate(bytes, 0, bytes.Length);
        byte[] hashBytes = new byte[keccak256.GetDigestSize()];
        // fill hash bytes
        keccak256.DoFinal(hashBytes, 0);
        return hashBytes;
    }
```

### The focus is:
```csharp
// Keccak-256 Algorithm
KeccakDigest keccak256 = new(256);

// size of digest (256 bits - 32bytes) as size of array
byte[] hashBytes = new byte[keccak256.GetDigestSize()];
```